File: packages/antgrid_relay_client/lib/src/flow.dart
/// Flow-control constants for the sealed session, hand-mirrored from
/// `packages/antgrid-wire/src/flow.ts`. Keep the two in lockstep by value: a
/// sender's window and its peer's credit batch only agree because both ends
/// compile the same numbers, and nothing on the wire negotiates them.
library;

/// Sealed bytes a sender may have in flight PER CHANNEL beyond what the peer
/// has credited. Holds one maximal sealed fragment with room to pipeline the
/// next as credits arrive, and is small enough that a control frame written
/// behind a full preview window waits only this much link time.
const int kChannelWindowBytes = 2097152;

/// Sealed bytes a sender may have in flight PER SOCKET, both channels together.
/// Bounds what a liveness frame written now sits behind, since neither client
/// can read its socket buffer — `dart:io` exposes nothing, and Bun's client
/// `bufferedAmount` reads 0 however much is queued. One window plus 1 MiB, so
/// control still flows while preview is full.
const int kSocketInflightBytes = 3145728;

/// A receiver credits a channel once this many uncredited bytes have arrived on
/// it. A quarter of the window, so the sender keeps three quarters of a window
/// of headroom between credits.
const int kCreditBatchBytes = 524288;

/// How old a credit-time anchor must be before bytes it saw written, and the
/// peer has still not counted, are presumed lost. Two liveness ticks: the peer
/// credits at least once per tick and the relay delivers a channel in order, so
/// a credit generated this long after a write has counted it if it ever
/// arrived. Time rather than a credit count, because byte-triggered credits
/// land milliseconds apart on a fast link.
const int kWindowResyncAgeMs = 40000;

/// Per-channel cap on plaintext bytes waiting in the send queue. A message that
/// would push past it is dropped whole, since a fragment set is never split.
/// Two maximal transfers: one queued behind another.
const int kMaxSendQueueBytes = 67108864;

/// nonce(12) + GCM tag(16). A sealed frame is this much longer than its utf8
/// plaintext, which is what lets a sender check its window before sealing.
const int kSealOverheadBytes = 28;

/// A channel blocked at the send gate for this long with data queued is logged
/// once, so a peer that stops crediting reads as "alive but not crediting"
/// rather than as a dead socket.
const int kWindowStallWarnMs = 5000;

File: packages/antgrid-wire/src/flow.ts
/** Per-channel flow control between the two ends of a peer connection. The Dart
 *  client hand-mirrors these in `packages/antgrid_relay_client/lib/src/flow.dart`;
 *  the two must stay in lockstep. */

/** Plaintext bytes a sender may have in flight PER CHANNEL beyond what the peer
 *  has credited. Holds one max-size fragment (1.4 MB) with room to pipeline
 *  the next one as credits arrive; small enough that a control frame written
 *  behind a full preview window waits ≤ 2 MiB of link time. */
export const CHANNEL_WINDOW_BYTES = 2_097_152;

/** Plaintext bytes a sender may have in flight across both channels on one peer
 * connection. Window + 1 MiB leaves headroom for control while preview is full. */
export const SOCKET_INFLIGHT_BYTES = 3_145_728;

/** Receiver credits once this many uncredited bytes arrived on a channel.
 *  window/4: the sender keeps ≥ 3/4 window of headroom between credits. */
export const CREDIT_BATCH_BYTES = 524_288;

/** How old a credit-time anchor must be before bytes it saw written, and the
 *  peer has still not counted, are presumed lost. Two liveness ticks: the peer
 *  credits at least once per tick and the the native record stream is ordered,
 *  so a credit generated this long after a write has counted it if it ever
 *  arrived. Time rather than a credit count, because byte-triggered credits
 *  land milliseconds apart on a fast link. */
export const WINDOW_RESYNC_AGE_MS = 40_000;

/** Per-channel cap on plaintext bytes waiting in the send queue. A message
 *  that would push past it is dropped whole (fragment sets are never split).
 *  2 × MAX_TRANSFER_BYTES: one maximal transfer queued behind another. */
export const MAX_SEND_QUEUE_BYTES = 67_108_864;

/** A channel gate-blocked for this long with data queued is logged once. */
export const WINDOW_STALL_WARN_MS = 5_000;
